x-pack/plugins/ml/public/explorer/explorer_utils.js

```diff
@@ -0,0 +1,490 @@
+  const selectedJobs = jobs.filter(job => job.selected);
+  return selectedJobs.map(job => job.id);
+}
+
+export function selectedJobsHaveInfluencers(selectedJobs = []) {
+  let hasInfluencers = false;
+  selectedJobs.forEach((selectedJob) => {
+    const job = mlJobService.getJob(selectedJob.id);
+    let influencers = [];
+    if (job !== undefined) {
+      influencers = job.analysis_config.influencers || [];
+    }
+    hasInfluencers = hasInfluencers || influencers.length > 0;
+  });
+  return hasInfluencers;
+}
+
+export function getSelectedJobs(previousSelectedJobs, jobs, selectedIds, appState) {
+  let previousSelectedCount = 0;
+  if (previousSelectedJobs !== null) {
+    previousSelectedCount = previousSelectedJobs.length;
+  }
+
+  // update the jobs' selected flag
+  const selectedJobs = [];
+  jobs.forEach((job) => {
+    job.selected = (indexOf(selectedIds, job.id) !== -1);
+    if (job.selected) {
+      selectedJobs.push(job);
+    }
+  });
+
+  // Clear viewBy from the state if we are moving from single
+  // to multi selection, or vice-versa.
+  appState.fetch();
+  if (
+    (previousSelectedCount <= 1 && selectedJobs.length > 1) ||
+    (selectedJobs.length === 1 && previousSelectedCount > 1)
+  ) {
+    delete appState.mlExplorerSwimlane.viewBy;
+  }
+  appState.save();
+
+  return new Promise((resolve) => {
+    // Populate the map of jobs / detectors / field formatters for the selected IDs.
+    mlFieldFormatService.populateFormats(selectedIds, getIndexPatterns())
+      .catch((err) => {
+        console.log('Error populating field formats:', err);
+        resolve(selectedJobs);
+      })
+      .then(() => {
+        resolve(selectedJobs);
+      });
+  });
+}
+
+export function getSelectionTimeRange(selectedCells, interval) {
+  // Returns the time range of the cell(s) currently selected in the swimlane.
+  // If no cell(s) are currently selected, returns the dashboard time range.
+  const bounds = timefilter.getActiveBounds();
+  let earliestMs = bounds.min.valueOf();
+  let latestMs = bounds.max.valueOf();
+
+  if (selectedCells !== null && selectedCells.times !== undefined) {
+    // time property of the cell data is an array, with the elements being
+    // the start times of the first and last cell selected.
+    earliestMs = (selectedCells.times[0] !== undefined) ? selectedCells.times[0] * 1000 : bounds.min.valueOf();
+    latestMs = bounds.max.valueOf();
+    if (selectedCells.times[1] !== undefined) {
+      // Subtract 1 ms so search does not include start of next bucket.
+      latestMs = ((selectedCells.times[1] + interval) * 1000) - 1;
+    }
+  }
+
+  return { earliestMs, latestMs };
+}
+
+export function getSelectionInfluencers(selectedCells, fieldName) {
+  const influencers = [];
+
+  if (
+    selectedCells !== null &&
+    selectedCells.fieldName !== undefined &&
+    selectedCells.fieldName !== VIEW_BY_JOB_LABEL
+  ) {
+    selectedCells.lanes.forEach((laneLabel) =>{
```

Comment:
nit, could be rewritten using a map.
not important.

Reply:
Done in 7eb77f19ab.